### DISCUSSION TITLE
Recommendations for audio editing (trim/cut)

### DISCUSSION THREAD

**Discussion by bernardwiesner** [katspaugh/wavesurfer.js#3557] [Q&A] [answered]
I am investigating the options for audio editing with wavesurfer. I came across this issue: https://github.com/katspaugh/wavesurfer.js/issues/419, but its almost a decade old, so I am wonder if there are any alternatives/simpler ways of editing, especially trimming and cutting using wavesurfer.

I would especially appreciate a feedback from @katspaugh on this, since you wrote this a decade ago: https://stackoverflow.com/questions/24551854/cut-and-paste-audio-using-web-audio-api-and-wavesurfer-js, but anyone else's feedback is also much appreciated.

How would you implement cropping and trimming bits of audio nowaday's if you would need to? Is the answer above still the best/simplest solution?

Thanks!

**Comment by katspaugh** [comment, marked as answer]
I think #419 still makes sense.

Basically the steps would be:

1. Set a higher `sampleRate` in[ wavesurfer options](https://wavesurfer.xyz/docs/types/wavesurfer.WaveSurferOptions) so that the decoded data is in audio quality
2. Get the decoded data using [wavesurfer.getDecodedData](https://wavesurfer.xyz/docs/classes/wavesurfer.default#getDecodedData) on ready
3. [Slice](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/TypedArray/slice) the decoded arrays
4. Turn the array back into audio – that's the tricky part
  * You could add a WAV header as per https://github.com/katspaugh/wavesurfer.js/issues/419#issuecomment-265530107
  * Or encode the audio using the new [AudioEncoder API](https://developer.mozilla.org/en-US/docs/Web/API/AudioEncoder)
  * Or use some WebAssembly encoder

**Reply by bernardwiesner** [reply]
Thanks for the reply!

What do you think about using ffmpeg in browser to trim/cut and pass the edited audio to wavesurfer? https://github.com/ffmpegwasm/ffmpeg.wasm
Could you think of cons of using ffmpeg in browser over the above approach?

**Reply by katspaugh** [reply]
Other than being 10 mb it's a great solution!

**Reply by bernardwiesner** [reply]
thanks, appreciate the feedback!

**Reply by bernardwiesner** [reply]
Hi @katspaugh 
Do you know of any way of improving the performance and memory usage of `audioCtx.decodeAudioData(audioData)` ? 

So everything is working great with ffmpeg.wasm, I even got the download size down to 7mb by using a custom build, but the bottleneck now is decoding the audio data to generate the new peaks when editing.

`decodeAudioData` is consuming a lot of memory and cpu, which makes every operation (cut/edit/etc) expensive. 

Is there some way you can think to generate the waveform with less accurate peaks, and reduce memory/cpu consumption either with decodeAudioData or with some other approach?

I know I can just cut the waveform as well, but I also want the ability to add new audio pieces to the waveform, and it would be easier to just generate all peaks again on every operation, even if the peaks are less accurate.

**Reply by katspaugh** [reply]
Since you already have ffmpeg, you could use that to decode the audio and pass this data to wavesurfer.

```js
wavesurfer.load(url, peaks, duration)
```

**Reply by bernardwiesner** [reply]
you mean generate the peaks with ffmpeg? 

I tried investigating how to do that, but it does not seem possible as far as I understand. The closes I could find is using ffprobe instead of ffmpeg [here](https://stackoverflow.com/questions/32254818/generating-a-waveform-using-ffmpeg/32276471#32276471). Unfortunately ffprobe is not part of the ffmpeg.wasm library. 

Or do you mean decode the audio in another way using ffmpeg?

**Reply by katspaugh** [reply]
With CLI it would be something like

```
ffmpeg -i input_file -f f32le -acodec pcm_f32le -ac 1 -ar 44100 pipe:1 | od -v -An -tf4 > output.txt
```

**Reply by bernardwiesner** [reply]
that works really well on the cli! But no idea how to do the `od -v -An -tf4` part with JS... is it possible?

**Reply by bernardwiesner** [reply]
Unfortunately I tested using web-assembly (ffmpeg) decoding the audio with above command and the performance was worse than using AudioContext. Memory usage was less, but processing time was much longer.

I think I will have to implement manually cutting the waveform using the sample rate (peaks.length / duration) in order to avoid decoding the audio every time.
```js
peaks.slice(start * sampleRate, end * sampleRate);
```

**Reply by bernardwiesner** [reply]
Hi @katspaugh 
I have a follow up question on your initial suggestion. You suggested I could use a WebAssembly encoder, do you have any suggestion for one? I tried with ffmpeg but does not seem to support that.

Thanks!

**Reply by katspaugh** [reply]
I haven't tried any. The first google result: https://github.com/arseneyr/wasm-media-encoders

**Comment by Mattnificent** [comment]
I was able to implement this super easily with the help from the link by incorporating a second hidden wavesurfer to deal with the encoding challenge:
````
let wavesurfer = WaveSurfer.create({
    container: '#wave_container_' + row.i,
    waveColor: '#4F4A85',
    progressColor: '#383351'
});

if (isTrimmed) {

    let trim_calc_wavesurfer = WaveSurfer.create({
        container: '#tcwc_' + row.i,
        waveColor: '#4F4A85',
        progressColor: '#383351'
    });

    GetAudioClipBlob(row.i)
        .then((blob) => {
            trim_calc_wavesurfer.loadBlob(blob);
        });

    /** When the audio is both decoded and can play */
    trim_calc_wavesurfer.on('ready', () => {
        console.log('Ready', trim_calc_wavesurfer);

        let fullDuration = trim_calc_wavesurfer.backend.getDuration();
        //let roundDuration = Math.round(fullDuration * 1000) / 1000;

        let startTime = row.s ?? 0;
        let endTime = row.f ?? fullDuration;
        let trimmedDuration = endTime - startTime;

        let originalBuffer = trim_calc_wavesurfer.backend.buffer;
        let startBufferIndex = startTime * originalBuffer.sampleRate;
        let endBufferIndex = endTime * originalBuffer.sampleRate;
        let trimmedBufferLength = trimmedDuration * originalBuffer.sampleRate;


        // Source: https://stackoverflow.com/a/24596314/2413712
        let copiedBuffer = trim_calc_wavesurfer.backend.ac.createBuffer(
            originalBuffer.numberOfChannels,
            trimmedBufferLength,
            originalBuffer.sampleRate
        );

        for (var i = 0; i < originalBuffer.numberOfChannels; i++) {
            let originalChanData = originalBuffer.getChannelData(i);
            let copiedChanData = copiedBuffer.getChannelData(i);

            for (var j = startBufferIndex, k = 0; j < endBufferIndex; j++, k++) {
                copiedChanData[k] = originalChanData[j];
            }
        }

        wavesurfer.loadDecodedBuffer(copiedBuffer);
    });
}
else {

    GetAudioClipBlob(row.i)
        .then((blob) => {
            wavesurfer.loadBlob(blob);
        });
}
````

This library is brilliant - I am excited to see what else I can do with it!
